New article: Private network access update (#518)
GoogleChrome/developer.chrome.com: site/en/blog/private-network-access-update/index.md

---
layout: 'layouts/blog-post.njk'
title: Private Network Access (CORS-RFC1918) updates
authors:
  - agektmr
  - titouan

> description: TBD

Reviewer: Please add the description here.

* Recommendations for mitigating the impact of the changes:
    * Receive reports when there are unexpected non-secure private network
      requests.
    * Avoid crossing address spaces.  
    * Serve embedder page and embedded content over HTTPS.

## What is Private Network Access

[Private Network Access](https://wicg.github.io/private-network-access/)
(formerly known as CORS-RFC1918) restricts the ability of websites to send
requests to servers on the private network. According to the specification, such
requests are only allowed from secure contexts. In addition, the specification
extends the Cross-Origin Resource Sharing (CORS) protocol so that websites now
have to explicitly request a grant from servers on the private network before
being allowed to send arbitrary requests.

**Private network requests** are requests whose target server's IP address is
more private than that from which the request initiator was fetched. For
example, a request from a public website (`https://example.com`) to a private
website (`http://router.local`), or a request from a private website to
localhost.

{% Img src="image/YLflGBAPWecgtKJLqCJHSzHqe2J2/nSGfU9xMJxIy6lMjJiMx.png",

> alt="ALT_TEXT_HERE", width="800", height="512" %}

Reviewer: Missing alt description here.

this deprecated feature in production before enforcement is enabled.

Starting in Chrome 92, private network requests initiated from non-secure
contexts will be blocked by Chrome. No request will be sent. Instead, an error
message will be logged in the DevTools console. For security reasons, this
manifests as a TypeError for fetch requests from Javascript.

## Recommended developer actions

If you know you have a website that fetches subresources from a private network,
here are instructions on how you can mitigate the impact of the upcoming
deprecation.

### Receive reports when there are unexpected non-secure private network requests {: #receive-reports}

We highly recommend that you set up a reporting endpoint server and send a
`Reporting-Endpoint` header on your website's documents to keep track of
unexpected non-secure private network requests. This will also serve to warn you
of other upcoming deprecations and errors your clients encounter in the wild.

To receive reports, there are a couple of known SaaS solutions.

* [https://report-uri.com](https://report-uri.com)
* [https://uriports.com](https://uriports.com)

{% Aside %}

> [Please let us know](https://github.com/GoogleChrome/web.dev/issues) if you know
> any other options so we can add them to the list.
> 
> We are not aware of any open source solutions that can accumulate reports in a
> database. We welcome to list them here as well if you ever build one.

Reviewer: ```suggestion
We are not aware of any open source solutions that can accumulate reports in a
database. If you know any other solutions, especially open source ones, [let us know](https://github.com/GoogleChrome/web.dev/issues) so we can add them to the list.
```

{% endAside %}

Once the endpoint is set up, send the endpoint URL with a `Reporting-Endpoint`
header with the top-level document.

```http
Reporting-Endpoints: default="https://example.com/reports"
Report-To: { group: 'default', max_age: 86400, endpoints: [{ url: 'https://reporting-endpoint.glitch.me/post'}]}
```

{% Aside %}

The Reporting API is undergoing transition to [a new
version](https://w3c.github.io/reporting/) (from `Reporting-To` to
`Reporting-Endpoint`). Chrome is planning to release it soon, but will leave the
older API in place for some time. Firefox is also [considering the new
API](https://bugzilla.mozilla.org/show_bug.cgi?id=1620573). You may want to use
both APIs during the transition.

{% endAside %}

### Avoid crossing address spaces

The simplest way around the issues caused by private network requests is to

> avoid straddling the boundary between public and private IP addresses. This can
> be achieved in one of two ways:

Reviewer: ```suggestion
avoid straddling the boundary between public and private IP addresses. There are
two ways to achieve this.
```

to routers filtering out private IP addresses from DNS responses. Please reach
out if you find yourself in this situation, Chrome would love to know more about
your use case and discuss available options.

## Plans for the future

Restricting private network requests to secure contexts is only the first step
in launching Private Network Access. Chrome is working towards implementing the
rest of the specification in the coming months. Stay tuned for updates!

### CORS preflight requests

The second part of Private Network Access is to gate private network requests
initiated from *secure* contexts with CORS preflight requests. The idea is that
even when the request was initiated from a secure context, the target server is
asked to provide an explicit grant to the initiator. The request is only sent if
the grant is successful.

In short, a CORS preflight request is an HTTP `OPTIONS` request carrying some
`Access-Control-Request-*` headers indicating the nature of the subsequent
request. The server can then decide whether or not to grant fine-grained access
by responding `200 OK` with `Access-Control-Allow-*` headers.

> More details on this can be found in [the
> specification](https://wicg.github.io/private-network-access).

Reviewer: ```suggestion
Find more details about this in the
specification](https://wicg.github.io/private-network-access).
```
